% OUTLINE:

% 1. Introduction to problem
%   a. Viruses are currently being classified (identification of viral host) 
%   based on a set of 
%   criteria specified by a committee of experts.
%   b. Propose to develop a more systematic, interpretable model
%   for classifying viruses based on viral genomic data.
%   c. This allows for fast classification of new viruses, particularly
%   viruses that threaten to cause pandemics. This would also allow
%   us to learn highly conserved regions/subsequences of viral genomes that are 
%   strong predictors of viral hosts.

% 2. Building the feature space + Algorithm
Motivated by the framework in Leslie et al. \cite{leslie}, we represent each viral protein
sequence in a feature space spanned by the set of all possible subsequences of amino acids
of a fixed length $k$. Each protein is then represented by counts of ($k$,$m$)-patterns
in its sequence, where the ($k$,$m$) pattern generated by a subsequence of length $k$ 
contains all $k$-length subsequences differing from it by at most $m$ mismatches. Non-zero
values for $m$ allow us to better capture rapidly mutating, yet conserved, viral subsequences
(or genomic regions) that are informative of the host of the virus.
%   c. Each protein is also assigned one of 3 labels {invertebrate, plant, vertebrate}. 
%   d. Armed with this representation, we use multi-class Adaboost to learn sparse, interpretable 
%   Alternating Decision Trees to predict the host of each virus (more specifically,
%   viral protein). The decision tree outputs a real-valued vector prediction (with
%   dimensionality equal to the number of classes) and the viral protein is assigned
%   to the class with the largest predicted output.

% 3. Results
%   a. We evaluate model accuracy at each round of boosting using a multi-class ROC score
%   b. We also represent the first 5 k-mers selected by boosting on each protein sequence (with
%   lengths scaled to [0,1]) - this illuminates regions of protein sequences that are conserved
%   in viruses of a specific host.

% 4. Next steps
%   a. Extend this framework to predict the hosts of viral families that are more difficult
%   to classify


