Classification of viruses into a taxonomic system based on phenotypic characteristics
such as morphology, mode of replication, host organism and type of disease caused is
a central problem in viral biology. Viruses are currently classified based on two major
schemes --- the ICTV\footnote{International Committee on Taxonomy of Viruses} system
and the Baltimore classification system, both of which have their own naming systems
and classification guidelines. Predicting the host organism of a newly discovered virus,
however, remains a challenging problem with different classification schemes having
different phenotypic criteria for each host class. Identifying the native 
host of viruses that threaten to cause pandemic outbreaks is critical to developing a 
functional understanding of the virus which could then help in designing appropriate 
vaccines and other countermeasures to control such outbreaks. This work proposes to 
develop a more systematic, 
interpretable model for predicting virus hosts based on patterns in viral protein sequence data. 
An interpretable model would also help identify highly conserved, functional regions (or subsequences) 
of viral genomes that are strong predictors of viral hosts.

Motivated by the framework developed in Leslie et al. \cite{leslie}, we represent each viral protein
sequence in a mismatch feature space spanned by the set of all possible subsequences of amino acids
of a fixed length $k$. Each protein is then represented by counts of ($k$,$m$)--patterns
generated by all contiguous $k$-length subsequences in its sequence. The ($k$,$m$)--pattern 
generated by a string $s$ of length $k$, also called its $m$-neighborhood and denoted by 
$\mathcal{N}_{k,m}(s)$, is the set of all $k$-length strings with a Hamming distance of
at most $m$ from it. Non-zero values for $m$ allow us to better capture rapidly mutating, 
yet functionally conserved, viral subsequences (or genomic regions) that are informative 
of the host of the virus. Each protein is also assigned one of 3 labels --- invertebrate, plant
or vertebrate --- depending on the host of the virus from which the protein was extracted.

Armed with this representation, we use multi-class Adaboost\cite{adaboost, multiclass} to learn sparse, 
interpretable models to predict the `host' of a viral protein, from which the host of the virus can then
be discerned. The model used is an Alternating Decision Tree\cite{adts} that outputs a real-valued vector 
prediction for each input protein --- the dimensionality of the vector prediction equals the number of 
classes. 
\begin{equation}
f(\mathbf{x}) = \sum_{l=1}^{L} \alpha_l \mathbf{v}_l \phi_l(\mathbf{x})
\end{equation}
where $f : \mathbf{x} \rightarrow \mathbb{R}^3$, $L$ is the number of simple decision rules 
$\phi_l(.)$ in the model (nodes in the decision tree), $\alpha_l$ is a non-negative weight 
associated with each rule and $\mathbf{v} \in \{-1,1 \}^3$ is a sign vector associated with each 
rule. The viral protein is then assigned to the class associated with the index corresponding
to the largest value in the predicted output.

In this work, we learnt a predictive and interpretable model to identify hosts of viruses 
belonging to the family \emph{Picornaviridae} --- a family of viruses that contain a single 
stranded, positive sense RNA less than 10 kilobases in length. The accuracy of the alternating 
decision tree model, at each round of boosting, was evaluated using a multi-class ROC score.
At each point on the ROC curve, a protein is considered to be classified correctly if the 
real-valued output for the true class of that protein is greater than the threshold value
corresponding to that point on the ROC curve. We observe that the smoothing effect introduced
by using the mismatch feature space allows for improved prediction accuracy for larger
values of $m$. Further note that prediction accuracy reaches a maximum at $m=5$; this feature space is
optimal for the algorithm to learn predictive, conserved amino-acid subsequences from rapidly 
mutating viral protein sequences. To visualize the $k$-mer subsequences selected by 
Adaboost, we indicate occurences of ($k$,$m$)--patterns of each subsequence on viral protein 
sequences grouped by their label, with their lengths scaled to $[0,1]$ (see Fig. 2) - this helps 
elucidate regions of protein sequences that are conserved in viruses of a specific host. Encouraged 
by the success of this approach to classify picornaviruses, future work will involve extending 
this approach to more difficult-to-classify virus families.

\begin{figure}
\begin{minipage}[t]{0.5\linewidth}
\incgraph{plot_auc_12.pdf}
\caption{\small{Plot of test AUC as a function of boosting round for 12-mers. Note that the test AUC decreases
after $m=5$ --- the optimal $m$-neighborhood necessary to capture predictive, conserved regions
in the viral genome.}}
\label{auc}
\end{minipage}\quad
\begin{minipage}[t]{0.5\linewidth}
\incgraph{kmer_visualization_9.pdf}
\caption{\small{A visualization of the relative location on the amino acid sequence of the 
($k$,$m$)--patterns of each of the first five 9-mers selected by Adaboost, with the viral proteins
vertically stacked and grouped by host. The selected 9-mers are listed on the right in corresponding colors.}}
\label{kmers}
\end{minipage}
\end{figure}


%Is there a biological reason to expect groups of viral genomic regions to be conserved
%as a group? If so, learning group sparsity structures among the features would be an 
%additional future work.
